packages/outstatic/src/pages/add-custom-field/index.tsx

```diff
@@ -0,0 +1,557 @@
+import camelCase from 'camelcase'
+import { useContext, useEffect, useState } from 'react'
+import { FormProvider, SubmitHandler, useForm } from 'react-hook-form'
+import capitalize from 'capitalize'
+import * as yup from 'yup'
+import { AdminLayout, Input } from '../../components'
+import Alert from '../../components/Alert'
+import Modal from '../../components/Modal'
+import { OutstaticContext } from '../../context'
+import { useCreateCommitMutation } from '../../graphql/generated'
+import { CustomField, CustomFields, customFieldTypes } from '../../types'
+import useFileQuery from '../../utils/useFileQuery'
+import useNavigationLock from '../../utils/useNavigationLock'
+import useOid from '../../utils/useOid'
+import { createCommit as createCommitApi } from '../../utils/createCommit'
+
+type AddCustomFieldProps = {
+  collection: string
+}
+
+type CustomFieldForm = CustomField & { name: string }
+
+export default function AddCustomField({ collection }: AddCustomFieldProps) {
+  const {
+    contentPath,
+    monorepoPath,
+    session,
+    repoSlug,
+    repoBranch,
+    repoOwner
+  } = useContext(OutstaticContext)
+  const [createCommit] = useCreateCommitMutation()
+  const fetchOid = useOid()
+  const [hasChanges, setHasChanges] = useState(false)
+  const [customFields, setCustomFields] = useState<CustomFields>({})
+  const yupSchema = yup.object().shape({
+    title: yup
+      .string()
+      .matches(/^[aA-zZ\s]+$/, 'Only alphabets are allowed for this field.')
+      .required('Custom field name is required.'),
+    type: yup.string().oneOf(['text', 'string']).required(),
+    description: yup.string(),
+    required: yup.boolean().required()
+  })
+  const [adding, setAdding] = useState(false)
+  const [error, setError] = useState('')
+  const methods = useForm<CustomFieldForm>({
+    mode: 'onChange',
+    resolver: yupResolver(yupSchema)
+  })
+  const [showAddModal, setShowAddModal] = useState(false)
+  const [showDeleteModal, setShowDeleteModal] = useState(false)
+  const [deleting, setDeleting] = useState(false)
+  const { data: schemaQueryData, loading } = useFileQuery({
+    file: `${collection}/schema.json`
+  })
+  const [selectedField, setSelectedField] = useState('')
+  const [fieldName, setFieldName] = useState('')
+
+  const capiHelper = async ({
+    customFields,
+    deleteField = false
+  }: {
+    customFields: any
+    deleteField?: boolean
+  }) => {
+    const oid = await fetchOid()
+    const customFieldsJSON = JSON.stringify(
+      {
+        title: collection,
+        type: 'object',
+        properties: { ...customFields }
+      },
+      null,
+      2
+    )
+
+    const capi = createCommitApi({
+      message: `feat(${collection}): ${
+        deleteField ? 'delete' : 'add'
+      } ${fieldName} field`,
+      owner: repoOwner || session?.user?.login || '',
+      oid: oid ?? '',
+      name: repoSlug,
+      branch: repoBranch
+    })
+
+    capi.replaceFile(
```

Comment:
If custom fields are removed, what is the expected behavior?

1. Those pieces of data persist and are deleted when a document is re-saved
2. Those fields are removed from all documents in the collection

I'm leaning towards 1 (current behavior) but it can cause issues if you change the type of a custom field such that it breaks the form controls in the UI. For example, taking a string and turning it into a date picker.

Reply:
This is a very good point. I think we can keep it as it is and then ship a version later where all the data related to a custom field is removed from all documents to avoid the potential for conflicting data types.

I'll definitely add an explanation about this to the docs.

Reply:
sgtm. I'm fine with leaving prior custom fields intact